### PR TITLE
185973374 - TalentLMS User Creation Bug Fix

### PR DESCRIPTION
**Pull request by dtgreiner**
## Description

Update TalentLMS user creation to use more strict passwords. Touched files also updated with rubocop enforced formatting changes.

## Type of change
- [x] Bug fix
- [ ] New feature (adds functionality)
- [x] Code clean-up / housekeeping
- [ ] Dependency update

## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (rubocop)
- [x] I have updated the documentation (or not applicable)
- [x] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

### CODE REVIEW

File: app/models/talentlms/facade.rb [hunk 1]
```diff
@@ -89,14 +88,35 @@ def course_url(user, course_id, redirect_url, logout_url)
...
+      loop do
+        break if p =~ /(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[^A-Za-z0-9])/
+
+        char_replace = (0...length).to_a.shuffle.take(4)
+        p[char_replace[0]] = ('a'..'z').to_a.sample unless p =~ /(?=.*[a-z])/
+        p[char_replace[1]] = ('A'..'Z').to_a.sample unless p =~ /(?=.*[A-Z])/
+        p[char_replace[2]] = ('0'..'9').to_a.sample unless p =~ /(?=.*[0-9])/
+        p[char_replace[3]] = (('#'..'&').to_a + (':'..'?').to_a).sample unless p =~ /(?=.*[^A-Za-z0-9])/
+      end
+      return p
```
Comment (eanders) [inline app/models/talentlms/facade.rb]: There's got to be a more readable way to do this.  I _think_ the goal is to get a password with at least one of each category.  Can we just do:
```suggestion
      p = SecureRandom.urlsafe_base64(length - 4)
      lower_letter = ('a'..'z').to_a.sample
      upper_letter = ('A'..'Z').to_a.sample
      number = ('0'..'9').to_a.sample
      special = (('#'..'&').to_a + (':'..'?').to_a).sample
     
      "#{p}#{lower_letter}#{upper_letter}#{number}#{special}".shuffle
```

File: app/models/talentlms/facade.rb [hunk 2]
```diff
@@ -33,7 +32,7 @@ def login(user)
     # @return [String] URL to redirect the user to to login
     def create_account(user)
       login = "#{ENV['RAILS_ENV']}_#{user.id + Integer(ENV.fetch('DEV_OFFSET', 0))}"
-      password = SecureRandom.hex(8)
+      password = random_password(16)
```
Comment (eanders) [inline app/models/talentlms/facade.rb:35]: with the default being 16 below, we can probably just do:
```suggestion
      password = random_password
```

Comment (eanders) [inline app/models/talentlms/facade.rb]: with the default being 16 below, we can probably just do:
```suggestion
      password = random_password
```

**Review by eanders — APPROVED** [review]
@dtgreiner looks great!  Thanks.  Just a couple small tweaks, let me know if you want to discuss.

File: app/models/talentlms/facade.rb (hunk 2 above)
Comment (dtgreiner) [reply app/models/talentlms/facade.rb]: this has been adjusted

File: app/models/talentlms/facade.rb (hunk 1 above)
Comment (dtgreiner) [reply app/models/talentlms/facade.rb]: this has been adjusted

File: app/models/talentlms/facade.rb [hunk 3]
```diff
@@ -89,14 +88,31 @@ def course_url(user, course_id, redirect_url, logout_url)
...
+    def random_password(length = 16)
+      raise ArgumentError 'Length must be at least 8' if length < 8
+
+      p = SecureRandom.urlsafe_base64(length - 4)
+      lower_letter = ('a'..'z').to_a.sample
+      upper_letter = ('A'..'Z').to_a.sample
+      number = ('0'..'9').to_a.sample
+      special = (('#'..'&').to_a + (':'..'?').to_a).sample
+
+      "#{p}#{lower_letter}#{upper_letter}#{number}#{special}".chars.shuffle.join
```
Comment (eanders) [inline app/models/talentlms/facade.rb:115]: Ah, good catch!  After talking about this, I was thinking 
```
[
  p,
  lower_letter,
  upper_letter,
  number,
  special,
].shuffle.join
```
would be another approach, but I think I like yours better.  Thanks!

**Pull request merged**